lib/types.nix

```diff
@@ -318,7 +318,7 @@ rec {
             def;
         attrOnly = attrsOf elemType;
       in mkOptionType rec {
-        name = "loaOf";
+        name = builtins.trace "types.loaOf is deprecated; use types.attrsOf instead" "loaOf";
```

Comment:
Yeah this isn't what we need. The warning needs to go in the `merge` function. There one has access to all definitions, which can then be filtered to lists only, and for each of them one can print a warning as I showed in my comment. would be even neater to include the file and line where the user needs to do this change.

I generally think our error messages suck too much. It's not too difficult to make them better, especially with all the info the module system gives access to.

Reply:
Ok, I'm working on it.

Reply:
I took me a while but I think this should do it:
```
trace: warning: In file <nixpkgs/nixos/modules/services/misc/nix-daemon.nix>
a list is being assigned to the option config.users.users.
This will soon be an error as type loaOf is deprecated.
See https://git.io/fj2zm for more information.
Do
  users.users =
    { nixbld1 = {...}; nixbld2 = {...}; nixbld3 = {...}; ...}
instead of
  users.users =
    [ { name = "nixbld1"; ...} { name = "nixbld2"; ...} { name = "nixbld3"; ...} ...]
```

Reply:
Uhm, no it doesn't work when the set is missing the `name` attribute.

Reply:
I mean, it does but it's not pretty.